gensync: address build issues

trachten opened the pull request:
#### Description
**gensync: update to version 2.0.4**
* updated code to version 2.0.4
* updated to compile from any macports directory, addressing [2](https://github.com/nislab/gensync/issues/2)
* add legacy portgroup setting, as per [2](https://github.com/nislab/gensync/issues/2)
* deleted redundant line (cmake.out_of_source   yes), as per [2](https://github.com/nislab/gensync/issues/2)

Closes: [2](https://github.com/nislab/gensync/issues/2)
<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
* macOS 13.0.1 22A400 arm64
* Xcode 14.3.1 14E300c
* Apple clang 14.0.3
* NTL 11.5.1_0

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

ryandesign commented on net/gensync/Portfile:
```diff
@@ -15,7 +22,7 @@ description             Gensync: a library for efficient synchronization of data
                         algorithms to sync data between two nodes in a network. \
                         These algorithms include IBLTs, CPISyncs, HashSyncs, Cuckoo Syncs, and more.
-
+set includedir ${prefix}
```
Why is this line needed? I see no other mention of this `includedir` variable in this Portfile.

ryandesign replied:
Here, you are setting a Tcl variable. Doing this does not pass it anywhere as an environment variable.

ryandesign replied:
I see that your CMakeLists.txt is trying to use a variable of that name [this way](https://github.com/nislab/gensync-macports/blob/0897f0f18765068fd8c54238de77fbc426b3051f/CMakeLists.txt#L124C1-L132):

```cmake
# Check if the "includedir" variable is already set
if(NOT DEFINED includedir)
    set(includedir "/opt/local" CACHE PATH "Path to the include directory")
endif()


# Users may need to update these paths
include_directories(${includedir}/include)
link_directories(${includedir}/lib)
link_directories(${includedir}/lib64)
```

So the variable is misnamed: it's not the include directory; it's the prefix, in which the include and lib directories exist. CMake already has a variable for that, which MacPorts sets for you: [`CMAKE_INSTALL_PREFIX`](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html). You should use that rather than invent new variables. I am not certain but you may not even need to set these include and link directories; they might already be set for you based on the prefix.

You can definitely delete the lib64 line. There never has been and probably never will be any lib64 directories on macOS.

ryandesign reviewed (changes requested):
> [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?

The commit message guidelines haven't been followed.

nathanstrahs commented on net/gensync/Portfile:
```diff
@@ -1,10 +1,17 @@
+compiler.cxx_standard   2011
 
-github.setup            nislab gensync 1.0.4 v
+github.setup            nislab gensync-macports 2.0.4 v
```
@ryandesign Thanks for reviewing our request! The reason for the separate repository is that the gensync-macports repository contains a different CMakeLists.txt than our main gensync repository. We did this on purpose because we do not need some of the functionality of the original CMakeLists.

trachten replied:
Just to be clear ... we have other repos that use the same gensync core code but have specialized utility (e.g., benchmarking).

barracuda156 replied:
> @ryandesign Thanks for reviewing our request! The reason for the separate repository is that the gensync-macports repository contains a different CMakeLists.txt than our main gensync repository. We did this on purpose because we do not need some of the functionality of the original CMakeLists.

@nathanstrahs If changes are minor, you could make a patch for CMakeLists. If changes are huge, then make `CMakeLists.txt.macports` file and add something like this to the portfile:
```
post-extract {
    delete ${worksrcpath}/CMakeLists.txt
    move ${worksrcpath}/CMakeLists.txt.macports ${worksrcpath}/CMakeLists.txt
}
And use a normal repo.

ryandesign replied:
> The reason for the separate repository is that the gensync-macports repository contains a different CMakeLists.txt than our main gensync repository.

Maintaining a separate codebase of your project just for MacPorts still seems like an utterly bizarre development decision to me. I don't understand why you want to bifurcate your user base in this way. Imagine we find a bug in gensync. Which of the two repos do we file the bug in? What if we want to search for an existing bug to see if a problem has already been reported? Now there are two issue trackers in which we need to search. What if we want to submit a fix? Which repo do we file the pull request in? How do you keep the code in the two repos synchronized? Why do you want to create all this additional work for yourselves?

The first sentence of the readme in that repo is:

> GenSync-macports is the extended version for MACOS.

So then it is not only for MacPorts? It is also for non-MacPorts users on macOS?

Looking at this CMakeLists.txt, I see that this software [requires C++11](https://github.com/nislab/gensync-macports/blob/0897f0f18765068fd8c54238de77fbc426b3051f/CMakeLists.txt#L3). This must be reflected in the Portfile via the line `compiler.cxx_standard 2011`.

nathanstrahs commented on net/gensync/Portfile:
```diff
@@ -1,10 +1,17 @@
 
 PortSystem              1.0
+
+name                    gensync
```
@ryandesign Thanks! We will address this...

reneeotten commented:
this has been open for more than half a year now. @trachten are you going to follow-up on @ryandesign his comments or do you want yo close this PR?

trachten commented:
> this has been open for more than half a year now. @trachten are you going to follow-up on @ryandesign his comments or do you want yo close this PR?

Sorry ... I believe we have addressed the issues and just need to update the repo.

pmetzger commented:
@trachten Would be cool to get this finished.

pmetzger commented:
@trachten I'd like to suggest that we close this for now, and that you submit a new pull request when you're ready to move forward.

pmetzger commented:
@trachten Closing this for now. Please feel free to reopen it whenever you have time to proceed.

trachten commented:
Sounds good ... and sorry for the delay.

> On Feb 6, 2024, at 10:21 AM, Perry E. Metzger ***@***.***> wrote:
> 
> 
> @trachten <https://github.com/trachten> Closing this for now. Please feel free to reopen it whenever you have time to proceed.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/macports/macports-ports/pull/19620#issuecomment-1930035663>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACM5LU6VZ22K47EWRN7I543YSJC75AVCNFSM6AAAAAA2XOHJSOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMZQGAZTKNRWGM>.
> You are receiving this because you were mentioned.
> 

—
Prof. Ari Trachtenberg
Electrical and Computer Engineering
Boston University
***@***.***

barracuda156 commented:
@trachten Please make a new PR once the issues are addressed, so that we have in working in Macports.

Chenxy517 commented:
> @trachten Please make a new PR once the issues are addressed, so that we have it working in Macports.

We have addressed the issue mentioned by @ryandesign in the [latest PR](https://github.com/macports/macports-ports/pull/21192). `gensync-lib` is now compatible with both MacOS and Linux. Thank you for the comments!
